orchagent/orchdaemon.cpp

```diff
@@ -105,6 +111,43 @@ OrchDaemon::~OrchDaemon()
     events_deinit_publisher(g_events_handle);
 }
 
+void OrchDaemon::popRingBuffer()
+{
+    SWSS_LOG_ENTER();
+
+    // make sure there is only one thread created to run popRingBuffer()
+    if (!gRingBuffer || gRingBuffer->threadCreated)
+        return;
+
+    gRingBuffer->threadCreated = true;
+    SWSS_LOG_NOTICE("OrchDaemon starts the popRingBuffer thread!");
+
+    while (!ring_thread_exited)
+    {
+        gRingBuffer->pause_thread();
```

Comment:
RingBufffer is a singleton share by 2 threads, however there is no multithread protect.
So when RingBufffer instance delete by main thread, the gRingBuffer here will become a dangling pointer.

Reply:
but the testcase is single-threaded

and in the orchdaemon destructor, we set ring_thread_exited as true to avoid delete gRingBuffer before detach the ring thread

Reply:
The test case may be failed by other reason.

The variable name ring_thread_exited suggests that the ring thread has exited, but this is not actually the case. When ring_thread_exited is true, the while loop might still be executing, and you cannot predict the ring thread or the main thread exist first.

These two threads are scheduled by the system. The system might suspend the ring thread for a few milliseconds at any point in the while loop. During that time, the main thread might exit and destruct the ring buffer.

Reply:
There is another possible reason for the UT issue:

The dtor of orchdaemon will access gRingBuffer, also dtor of RingBuffer class will delete the instance. which dtor will invoke first? if RingBuffer class destroy first, then gRingbuffer also will become a dangling pointer.

My suggestion is RingBuffer should not be a singleton and release in the dtor of RingBuffer class.

Reply:
sure, 
1. a try catch block is added for the while loop in ring thread, to cover the case ring buffer is released before ring thread is detached.
2. ring buffer is released by its own dtor
in orchdaemon dtor , the local variable `gRingBuffer` is only used in this line to indicate whether we has enabled gRingMode
```
if (gRingBuffer) {
        ring_thread_exited = true;
        ring_thread.detach();}
```
if it's not nullptr, it indicates we enabled gRingMode, so need to detach the thread and set  ring_thread_exited as true, nothing else, so should be good here

Reply:
When access a dangling pointer on linux system, it will trigger a SIGSEGV signal.
And as I know a signal is not a C++ exception. so, try-catch will not work.

My suggestion is:
1. continue use singleton, but in while loop, always use get() method to access singleton. and get() method need improve to handle multithread, also while loop need handle get() return null pointer case (when singleton destructued). which is really difficult and low performance.

2.Don't use singleton, create a global instance and manage by OrchDaemon.

Reply:
singleton has API Get() to access the global and unique ring buffer

gRingBuffer is just a variable that records what Get() returns

we use it to compare with nullptr to check whether ring is enabled, we can always use OrchRing::Get() to access the unique ring buffer

Reply:
What will happen when dtor of ring buffer called before popRingBuffer() thread exit? popRingBuffer() thread will cause SIGSEGV and make orchagent crash.

This kind crash is very difficult to find root cause.

Reply:
No one would call the dtor of ringbuffer, its protected and can't be called by others